io-sim/test/Test/Control/Monad/Class/MonadMVar.hs

```diff
@@ -0,0 +1,325 @@
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TupleSections         #-}
+
+module Test.Control.Monad.Class.MonadMVar where
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadFork
+import           Control.Monad.Class.MonadMVar
+import           Control.Monad.Class.MonadTime
+import           Control.Monad.Class.MonadTimer
+import           Data.Foldable (traverse_)
+import           Data.Functor (($>))
+import           Data.Maybe (isNothing)
+
+import           Control.Monad.IOSim
+
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.HUnit
+import           Test.Tasty.QuickCheck (testProperty)
+
+tests :: TestTree
+tests =
+    testGroup "Control.Monad.Class.MonadMVar"
+    [ testGroup "putMVar"
+      [ testProperty "fairness (IOSim)" prop_putMVar_fairness_sim
+      , testCase "blocks on a full MVar (IOSim)"
+        unit_putMVar_blocks_on_full_sim
+      , testCase "blocks on a full MVar (IO)"
+        unit_putMVar_blocks_on_full_io
+      ]
+    , testGroup "takeMVar"
+      [ testProperty "fairness (IOSim)" prop_takeMVar_fairness_sim
+      , testCase "blocks on an empty MVar (IOSim)"
+        unit_takeMVar_blocks_on_empty_sim
+      , testCase "blocks on an empty MVar (IO)"
+        unit_takeMVar_blocks_on_empty_io
+      ]
+    , testGroup "tryTakeMVar"
+      [ testCase "does not block on an empty MVar (IOSim)"
+        unit_tryTakeMVar_empty
+      , testCase "does not block on a full MVar (IOSim)"
+        unit_tryTakeMVar_full
+      , testCase "return value on an empty MVar (IOSim)"
+        unit_tryTakeMVar_return_empty_sim
+      , testCase "return value on an full MVar (IOSim)"
+        unit_tryTakeMVar_return_full_sim
+      ]
+    , testGroup "tryPutMVar"
+      [ testCase "does not block on an empty MVar (IOSim)"
+        unit_tryPutMVar_empty
+      , testCase "does not block on a full MVar (IOSim)"
+        unit_tryPutMVar_full
+      , testCase "return value on an empty MVar (IOSim)"
+        unit_tryPutMVar_return_empty_sim
+      , testCase "return value on an full MVar (IOSim)"
+        unit_tryPutMVar_return_full_sim
+      ]
+    , testGroup "isEmptyMVar"
+      [ testCase "empty MVar is empty"    unit_isEmptyMVar_empty_sim
+      , testCase "full MVar is not empty" unit_isEmptyMVar_full_sim
+      ]
+    ]
+
+
+--
+-- putMVar
+--
+
+-- | Check that 'takeMVar' is fair.  This is test is only designed for 'IOSim'
+-- as it relies on its thread scheduling and determinism.
+--
+prop_putMVar_fairness
```

Comment:
maybe drop the `prop_` prefix from this as it might be confusing. I am not sure what's the best convention

Reply:
Isn't it a property?

Reply:
Right, I'll change the name.

Reply:
Renamed with a suffix `_property`.